File: app/src/main/java/com/arnold/sleepmonitor/utils/MSensorManager.java
package com.arnold.sleepmonitor.utils;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import com.arnold.sleepmonitor.utils.MLog;

public class MSensorManager implements SensorEventListener{
    private SensorManager sensorManager;
    private Sensor lightSensor;
    private Sensor gameRotateSensor;
    private Sensor linearAccelerometer;

    private double lightValue;
    private double vectorX;
    private double vectorY;
    private double vectorZ;

    private double accX;
    private double accY;
    private double accZ;

    public MSensorManager(Context context) {
        sensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
        lightSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT);
        gameRotateSensor = sensorManager.getDefaultSensor(Sensor.TYPE_GAME_ROTATION_VECTOR);
        linearAccelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION);
    }

    public void startListener() {
        MLog.d("MSensorManager", "startListener");
        sensorManager.registerListener(this, lightSensor, SensorManager.SENSOR_DELAY_NORMAL);
        sensorManager.registerListener(this, gameRotateSensor, SensorManager.SENSOR_DELAY_NORMAL);
        sensorManager.registerListener(this, linearAccelerometer, SensorManager.SENSOR_DELAY_NORMAL);
    }

    public void stopListener() {
        MLog.d("MSensorManager", "stopListener");
        sensorManager.unregisterListener(this);
    }

    @Override
    public void onSensorChanged(SensorEvent sensorEvent) {
        if (sensorEvent.sensor.getType() == Sensor.TYPE_LIGHT) {
            lightValue = sensorEvent.values[0];
            MLog.d("MSensorManager", "Light Value: " + lightValue);
        }

        if (sensorEvent.sensor.getType() == Sensor.TYPE_GAME_ROTATION_VECTOR) {
            // x * sin(θ/2)
            vectorX = sensorEvent.values[0];
            vectorY = sensorEvent.values[1];
            vectorZ = sensorEvent.values[2];
            MLog.d("MSensorManager", "Vector Value: " + vectorX + ", " + vectorY + ", " + vectorZ);
        }

        if (sensorEvent.sensor.getType() == Sensor.TYPE_LINEAR_ACCELERATION) {
            accX = sensorEvent.values[0];
            accY = sensorEvent.values[1];
            accZ = sensorEvent.values[2];
            MLog.d("MSensorManager", "Acc Value: " + accX + ", " + accY + ", " + accZ);
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int i) {

    }

    public double getLightValue() {
        return lightValue;
    }

}

File: app/src/main/java/com/arnold/sleepmonitor/MSensorManager.kt
package com.arnold.sleepmonitor

import android.annotation.SuppressLint
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.util.Log

class MSensorManager(context: Context?) : SensorEventListener {
    private var sensorManager: SensorManager? = null
    private var lightSensor: Sensor? = null
    private var linearAccelerometer: Sensor? = null

    var lightValue = 0f
    var accelerometerX = 0f
    var accelerometerY = 0f
    var accelerometerZ = 0f

    init {
        if (context != null) {
            this.sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        }
        lightSensor = sensorManager?.getDefaultSensor(Sensor.TYPE_LIGHT)
        linearAccelerometer = sensorManager?.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)
    }

    fun startListener() {
        Log.d("SensorManager", "Starting listener.")
        sensorManager?.registerListener(this, lightSensor, SensorManager.SENSOR_DELAY_NORMAL)
        sensorManager?.registerListener(this, linearAccelerometer, SensorManager.SENSOR_DELAY_NORMAL)
    }

    fun stopListener() {
        Log.d("SensorManager", "Stopping listener.")
        sensorManager?.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent?) {
        if (event != null) {
            if (event.sensor.type == Sensor.TYPE_LIGHT) {
                lightValue = event.values[0]
                Log.d("SensorManager", "Light value: $lightValue")
            }
            if (event.sensor.type == Sensor.TYPE_LINEAR_ACCELERATION) {
                accelerometerX = event.values[0]
                accelerometerY = event.values[1]
                accelerometerZ = event.values[2]
                Log.d("SensorManager", "Accelerometer values: $accelerometerX, $accelerometerY, $accelerometerZ")
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        Log.i("SensorManager", "The accuracy of the sensor have changed.")
    }
}